pkg/auth/service.go

```diff
@@ -474,299 +473,275 @@ func (s *DBAuthService) ListEffectivePolicies(ctx context.Context, username stri
+			return nil, nil, err
+		}
+		for _, group := range groups {
+			path := model.KvUserToGroup(username, group.DisplayName)
+			m := model.UserData{}
+			_, err := s.store.GetMsg(ctx, model.PartitionKey, path, &m)
+			if err != nil && !errors.Is(err, kv.ErrNotFound) {
+				return nil, nil, err
+			}
+			if !errors.Is(err, kv.ErrNotFound) {
+				userGroups = append(userGroups, group)
+				count += 1
+			}
 		}
-		p := &model.Paginator{}
-		users := make([]*model.User, len(dbUsers))
-		for i := range dbUsers {
-			users[i] = model.ConvertUser(dbUsers[i])
+		if paginator.NextPageToken == "" {
+			hasMoreGroups = false
 		}
-		if len(users) == params.Amount+1 {
-			// we have more pages
-			users = users[0:params.Amount]
-			p.Amount = params.Amount
-			p.NextPageToken = users[len(users)-1].Username
-			return &res{users, p}, nil
+		params = &model.PaginationParams{
+			After: paginator.NextPageToken,
 		}
-		p.Amount = len(users)
-		return &res{users, p}, nil
-	})
+	}
+	if count == params.Amount {
+		nextPage = params.After
+	}
+	p := &model.Paginator{Amount: count, NextPageToken: nextPage}
+	return userGroups, p, nil
+}
 
-	if err != nil {
-		return nil, nil, err
+func (s *KVAuthService) ListGroupUsers(ctx context.Context, groupDisplayName string, params *model.PaginationParams) ([]*model.User, *model.Paginator, error) {
+	if _, err := s.GetGroup(ctx, groupDisplayName); err != nil {
+		return nil, nil, fmt.Errorf("%s: %w", groupDisplayName, err)
 	}
-	return result.(*res).users, result.(*res).paginator, nil
+	var policy model.UserData
+	userGroupKey := model.KvUserToGroup("", groupDisplayName)
+
+	msgs, paginator, err := s.ListKVPaged(ctx, (&policy).ProtoReflect().Type(), params, userGroupKey, true)
+	if msgs == nil {
+		return nil, paginator, err
+	}
+	return model.ConvertUsersDataList(msgs), paginator, err
 }
 
-func (s *DBAuthService) WritePolicy(ctx context.Context, policy *model.BasePolicy) error {
-	_, err := s.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		if err := model.ValidateAuthEntityID(policy.DisplayName); err != nil {
-			return nil, err
-		}
-		for _, stmt := range policy.Statement {
-			for _, action := range stmt.Action {
-				if err := model.ValidateActionName(action); err != nil {
-					return nil, err
-				}
-			}
-			if err := model.ValidateArn(stmt.Resource); err != nil {
-				return nil, err
-			}
-			if err := model.ValidateStatementEffect(stmt.Effect); err != nil {
-				return nil, err
+func (s *KVAuthService) WritePolicy(ctx context.Context, policy *model.BasePolicy) error {
+	if err := model.ValidateAuthEntityID(policy.DisplayName); err != nil {
+		return err
+	}
+	for _, stmt := range policy.Statement {
+		for _, action := range stmt.Action {
+			if err := model.ValidateActionName(action); err != nil {
+				return err
 			}
 		}
+		if err := model.ValidateArn(stmt.Resource); err != nil {
+			return err
+		}
+		if err := model.ValidateStatementEffect(stmt.Effect); err != nil {
+			return err
+		}
+	}
```

Comment:
Maybe extract validation to a common place?

Reply:
done